torch/_dynamo/variables/torch.py

```diff
@@ -1046,29 +952,38 @@ def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
             graph_checkpoint, checkpoint = tx.output.graph, tx.copy_graphstate()
 
             def speculate_branch(branch):
+                # NB: 0 is predicate
+                ix = 1 if branch else 2
                 try:
-                    # NB: 0 is predicate
-                    ix = 1 if branch else 2
-                    return old_speculate_subgraph(
-                        args[ix], operands, graph_checkpoint, checkpoint
+                    ret_val, ret_graph, ret_lifted_freevars = speculate_subgraph(
+                        tx, args[ix], operands, graph_checkpoint, checkpoint
+                    )
+                except Exception as e:
+                    raise UserError(UserErrorType.DYNAMIC_CONTROL_FLOW, str(e)) from e
```

Comment:
This codepath gets used by both training and export. Just to check, is UserError an export-specific thing, or does it work with training as well?

The thing I am concerned about is that speculate_subgraph can `raise Unsupported`, which would lead to a graph break in training. However, that exception would be caught by this general `Exception`.  Does this end up squashing the graph break in training?